psh/ls/ls.c

```diff
@@ -333,16 +362,32 @@ static int psh_ls_printmultiline(size_t nfiles)
 {
 	fileinfo_t *files = psh_ls_common.files;
 	size_t ncols, nrows, *colsz;
-	unsigned int row, col, idx;
+	unsigned int row, col, idx, restcnt = 0;
+	bool printrest = false;
 
 	if ((colsz = psh_ls_computerows(&nrows, &ncols, nfiles)) == NULL)
 		return -ENOMEM;
 
 	for (row = 0; row < nrows; row++) {
+		if (printrest) {
+			printrest = false;
+			psh_ls_printrest(&files[idx]);
+			restcnt++;
+			putchar('\n');
+			continue;
+		}
 		for (col = 0; col < ncols; col++) {
-			if ((idx = col * nrows + row) >= nfiles)
+			if ((idx = col * nrows + row - restcnt) >= nfiles) {
 				continue;
-			psh_ls_printfile(&files[idx], max(files[idx].namelen, min(colsz[col], psh_ls_common.ws.ws_col)));
+			}
+			if (files[idx].namelen < psh_ls_common.ws.ws_col) {
+				/* printing width has to be lower than window size to avoid extra newline */
+				psh_ls_printfile(&files[idx], max(files[idx].namelen, min(colsz[col], psh_ls_common.ws.ws_col - 1)));
+			}
+			else {
+				psh_ls_printfile(&files[idx], psh_ls_common.ws.ws_col - 1);
+				printrest = true;
+			}
```

Comment:
Before I dig deeper into those changes, I wonder if it is necessary to manually break the file names? I understand that at the moment we compute the width in a wrong way, so that we print 80 characters using `printf` and then a newline using `putchar`. When working with vga as a console we see "two" newlines - one printed by our algorithm and one added by the tty to move the carrier to a new line. With `CONSOLE=serial` it works almost fine, with an exception that we add one extra line at the end, because `nrows` is too big.

The question is, wouldn't it be better and simpler to let the tty break those lines for us? Maybe you did that this way in order to make the output look the same when using vga or serial as CONSOLE? If that is what we are aiming for, then I agree that this is a correct direction.

Reply:
I thought about it. I didn't want to change too many things here and had no idea how to compute rows, when it's done automatically. So I've tried to adjust to the current approach. I will look at it second time and let You know if there is better a way.